opencog/atoms/value/StreamValue.cc

```diff
@@ -32,4 +32,13 @@ bool StreamValue::operator==(const Value& other) const
 	return &other == this;
 }
 
+bool StreamValue::operator<(const Value& other) const
+{
+	if (STREAM_VALUE != other.get_type())
+		return STREAM_VALUE < other.get_type();
+
```

Comment:
If instead, you just add `if (this->get_type() != other.get_type())` to `Value.h` and `Value.cc` then you won't have to nickel-n-dime to death all the derived classes.

Reply:
I am probably missing something obvious but I can't see how that could work. I need the if to decide between 2 options one which is the same in every class and one which is different.

I could put the 1st option in the parent somehow but the decision if we should use that or the specialized one still needs to happen in the child classes. Because If I put the branch into Value it can't know what to do in case the types are equal.

Reply:
It will always be true, all of the time, that `STREAM_VALUE == this->get_type()` whenever `this` is a `StreamValue` class. Likewise, it will always be true that  `FLOAT_VALUE == this->get_type()` whenever `this is a `FloatValue`.  Thus, it is enough to provide a default implementation  in the base class, instead of cut-n-pasting nearly identical code in each derived class.

Reply:
I still don't get it. The code kinda looks like
```
If (this->get_type() != other.get_type())
    return this->get_type() < other.get_type();
else
    Do something custom for this type
```
I haven't explicitly used an else but maybe I should. Anyway it's true that the condition and the first option are the same for all Values but I can't just move out a part of the Code and if I try to move the whole block it fails because I can't know what the custom part is. Unless you want me to do a switch statement that checks the Type we have and then does something custom according to that. But that seems like bad design if the code in Value is responsible for other Types.

Reply:
Never use `else` if you can avoid it. Indentation is evil.

custom: but all of your code is just cut-n-paste, all of the "custom" parts are identical.  If/when some value actually does need to do anything actually custom, it can always implement this method. But right now, they are all cut-n-paste of one-another.

However, I am starting to question the need for an `opeator<()` for Values, anyway. What does this do? What's the point? What's it good for?  Consider an audio stream -- should    `operator<()` compare the time-stamps on the audio streams? should it compare the loudness of the audio? Should it compare the URL of the audio source, ordered lexicographically? What, exactly, are we trying to accomplish, here?

Reply:
Well in this case the custom part looks like:
```
const StreamValue* sov = (const StreamValue*) &other;
return _value < sov->_value;
```
But since _value is not define in Value.h but instead in each of it's sub-classes I can't move that part into Value.h

Well, the reason I originally added this was so that I could use Values in an ordered container. For that, the ordering doesn't really need a semantic meaning just like the ordering of Atoms has no meaning. It's just an added benefit that for some Values like StringValue or FloatValue it does have a meaning.
But I mean we don't have to add this now while nobody is using it. It just seemed like a useful addition overall.

Reply:
Hmm. philosophically speaking "It just seemed like a useful" is considered to be a dangerous, bad coding habit. It is something that junior programmers love to do, and senior programmers (and project managers) detest.  Here's why (and let me pretend this is a commercial setting, but its also true for open source): It's not only code-bloat, its pernicious: three years from now, someone realizes that they need some kind of ordering that is different from what the current `operator<()` does. They add that. They might spend a lot of time trying to retrofit the existing `operator<()` be backwards-compat.  If no one is actually using `operator<()`, then this is time-wasted on the part of the developer.  Worse, some customer might discover this, and start using it (probably for some spurious, totally bogus reason) but "the customer is always right" so the project manager has to assign someone to write better unit tests, better integration tests, and check system test coverage. And if the customer gets borken, deal with that.   Meanwhile the system architect gets pulled in and proclaims "this is stupid, we need to rip this out". So, now we have to figure out how to deprecate the function, while maintaining backwards-compat for N years.

What has happened is that an idle afternoon of a junior developer has snowballed into ten-person-days, or more, of headache-inducing fixing, tracking, refactoring and redesign work down the road -- all to support something that really wasn't really needed, to begin with. The general moral of the story:  don't just add spurious code to a code-base just because it seems cute.  There really should be a strong reason for it: an immediate need or an obviously required feature.

Reply:
Well then I'll just get rid of these changes but it would have been nice if you could have told me that earlier. Because exactly for that reason I wrote in my initial message that these changes are currently not used by anything. I thought it was useful but I did want confirmation that adding code that's not used is okay. Guess next time I'll ask more directly.